src/collapse/dtrace.rs

```diff
@@ -213,6 +213,16 @@ impl Folder {
                     };
                     while let Some(data) = rx_input.recv().unwrap() {
                         if let Err(e) = folder.collapse_single_threaded(&data[..]) {
+                            // Use `try_send` here because we only need to send one
+                            // error back to the main thread for propogation (even if
+                            // multiple threads fail). If multiple threads fail,
+                            // the first one to do so will fill up our output channel
+                            // (which only has one slot); so plain `send` would block
+                            // here, which can't happen because we need each thread to
+                            // continuously pull values off the input channel or else
+                            // it may fill up and clog. This is also why we continue
+                            // after trying to send the error - to ensure each thread
+                            // is always pulling values off the bounded input channel.
```

Comment:
Reading through this, I wonder if we want to _not_ call `collapse_single_threaded` any more once we've produced an error. There's no reason to do any more processing at that point, right?

Reply:
I was going to do that actually (stop doing additional work if an error occurs), but thought you might balk at the added code / complexity. 

The cost of *not* doing it is just that the CPU does a little extra work and it takes a little bit longer for the overall collapse method to error. 

What do you think about the below?

```rust
let mut do_work = true;
while let Some(data) = rx_input.recv().unwrap() {
    if do_work {
        if let Err(e) = folder.collapse_single_threaded(&data[..]) {
            if let Err(e) = tx_output.try_send(Err(e)) {
                match e {
                    TrySendError::Full(_) => do_work = false,
                    TrySendError::Disconnected(_) => unreachable!(),
                 }
            }
     }
}

```

Reply:
Instead of using `do_work`, I would stick
```rust
let _ = rx_input.count();
break
```
inside the `if let Err(e) = ...` condition instead.

Reply:
I don't think we can break. Assume all threads error on receiving their first chunk of data, but the main thread still needs to send more. In that case there would be no thread to pull data off the input channel and the main thread would block forever trying to send more data on it.

Reply:
Nevermind - Totally disregard that. I misread your comment.

Reply:
I'm not sure what you mean by `count` (there's not `count` method on a sender). I just pushed something else that I think accomplishes what we want.

Reply:
I'm not sure what you mean by `count` (there's not a `count` method on `Receiver`). I just pushed something else that I think accomplishes what we want.

Reply:
From memory, `Receiver` implements `Iterator`, which has a `count` method. You may need to call `into_iter()` on it first.

Reply:
I just pushed some changes that restructure how we're doing work on the threads. Notably, this commit makes it so if any of the workers error they all stop doing unnecessary work and return. I only made these changes for dtrace, as soon I'll combine all the collapsers in more shareable code. Let me know what you think.